src/features/web-compat.js

```diff
@@ -43,6 +67,64 @@ export default class WebCompat extends ContentFeature {
         if (this.getFeatureSettingEnabled('presentation')) {
             this.presentationFix()
         }
+
+        if (this.getFeatureSettingEnabled('webShare')) {
+            this.shimWebShare()
+        }
+    }
+
+    /** Shim Web Share API in Android WebView */
+    shimWebShare () {
+        if (typeof navigator.canShare === 'function') return
+        this.defineProperty(Navigator.prototype, 'canShare', {
+            configurable: true,
+            enumerable: true,
+            writable: true,
+            value: canShare
+        })
+
+        if (typeof navigator.share === 'function') return
+        this.defineProperty(Navigator.prototype, 'share', {
+            configurable: true,
+            enumerable: true,
+            writable: true,
+            value: (data) => {
+                if (!canShare(data)) return Promise.reject(new TypeError('Invalid share data'))
+                if (this.#activeShare) {
+                    return Promise.reject(new DOMException('Share already in progress', 'InvalidStateError'))
+                }
+                if (!navigator.userActivation.isActive) {
+                    return Promise.reject(new DOMException('Share must be initiated by a user gesture', 'InvalidStateError'))
+                }
+
+                const dataToSend = {}
+                for (const key of ['title', 'text', 'url']) {
+                    if (key in data) dataToSend[key] = data[key]
+                }
+                if ('url' in data) dataToSend.url = new URL(data.url) // clean url
+
+                // eslint-disable-next-line promise/prefer-await-to-then
```

Comment:
Why not use await?

Reply:
good question. I wanted the return value be the same as #activeShare, but not sure why.. I'll change

Reply:
actually on the second thought I do like it this way because it makes a encapsulates #activeShare cleanup in itself. And error handling is kinda nicer then a try..catch, so I'd rather keep it, but let me know if you think it's better another way.

Reply:
One of many variants:

```javascript
                this.#activeShare = this.messaging.request(MSG_WEB_SHARE, dataToSend)
                try {
                      const resp = await this.#activeShare
                      this.#activeShare = null
                      if (resp.failure) {
                          switch (resp.failure.name) {
                          case 'AbortError':
                          case 'NotAllowedError':
                          case 'DataError':
                              throw new DOMException(resp.failure.message, resp.failure.name)
                          default:
                              throw new DOMException(resp.failure.message, 'DataError')
                          }
                      }
                      return resp
                } catch (err) {
                        this.#activeShare = null
                        throw new DOMException(err.message, 'DataError')
                }
```

Or:
```javascript
                this.#activeShare = this.messaging.request(MSG_WEB_SHARE, dataToSend)
                let resp
                try {
                        resp = await this.#activeShare
                } catch (err) {
                        this.#activeShare = null
                        throw new DOMException(err.message, 'DataError')
                }
                this.#activeShare = null
                if (resp.failure) {
                    switch (resp.failure.name) {
                    case 'AbortError':
                    case 'NotAllowedError':
                    case 'DataError':
                        throw new DOMException(resp.failure.message, resp.failure.name)
                    default:
                        throw new DOMException(resp.failure.message, 'DataError')
                    }
                }
                return resp
```

Or if you want the encapsulation (I think you're on about):

```javascript
            async function createWebShare (dataToSend) {
                try {
                      const resp = await this.messaging.request(MSG_WEB_SHARE, dataToSend)
                      if (resp.failure) {
                          switch (resp.failure.name) {
                          case 'AbortError':
                          case 'NotAllowedError':
                          case 'DataError':
                              throw new DOMException(resp.failure.message, resp.failure.name)
                          default:
                              throw new DOMException(resp.failure.message, 'DataError')
                          }
                      }
                      return resp
                } catch (err) {
                        throw new DOMException(err.message, 'DataError')
                }
            }
```

And then in this method:
```javascript
try {
    this.#activeShare = await createWebShare(dataToSend)
    this.#activeShare = null
} catch {
    this.#activeShare = null
}
```